### PR TITLE
dark mode toggle, modularization into Svelte components, filter "and/or" mode

### LINKED ISSUE #141: [FEATURE] Add toggle to allow user to search tags with OR logic or AND logic
**Describe your suggested improvement**
Right now, I can only search for one particular topic at a time. I want to change the search logic for tags to represent searching for tag1 or tag2 instead of tag1 and tag2.

**Describe the benefit**
For instance, I tried to search for Article + Climate Tech and nothing came up -- but I think it would be helpful to have that be represented as Articles or Climate Tech resources so that the user is more likely to see results.

**Would you be capable/willing to implement the improvement?**
Yes

**Additional comments**
Please let me know if this is something you would want me to change! I'll start working on the change in the meantime. In the long run, I could add a more complex search that allows for both and/or logic and potentially separates tags into categories, etc.

CODE REFERENCE LOCATION: 
[filterResources Location](https://github.com/ClimateTown/knowledge-hub/blob/c39e88f008f7443a9b48987c2328cc41886d84b9/src/routes/resources/%2Bpage.svelte#LL26C26-L26C26)

### PR DESCRIPTION
**Pull request by dmlb**
## Describe your changes

- create component for dark mode toggle logic
    - update tailwind dark mode config to be class-based
    - check os settings when initializing and save user selection in local storage
    - toggle classes for dark mode
    - update tag dark mode CSS to use classes
- create component button/button links for consistent styles dark/hover/disable
- create component tag wrapper
- create component checkbox
- create component for back to top scroll button (and add to youtube page)
- create component for resource nav
- create component for Filter Form
- move functions into lib/utils
- move scrolling constants to lib/constants
- replace items with components in the resource and youtube page
- update the resources filter to do and/or based on user switch

Screen Shot 2023-07-13 at 8 14 54 PM
Screen Shot 2023-07-13 at 8 15 05 PM

## Related issue number/link
Dark mode button: https://github.com/ClimateTown/knowledge-hub/pull/282#issuecomment-1604546541
Refactoring: Contributes to #299 (previously #209)
Filter mode: Fixes #141

### CODE REVIEW

**Comment by VeckoTheGecko** [comment]
We use [Bootstrap Icons](https://icons.getbootstrap.com/icons/sun-fill/) for the icons (I'll update the README in another PR with this info).

Icons ideas:

- [Sun (light mode)](https://icons.getbootstrap.com/icons/sun-fill/)
- [Moon (dark mode)](https://icons.getbootstrap.com/icons/moon-stars-fill/)
- [Intersection (and)](https://icons.getbootstrap.com/icons/intersect/)
- [Union (or)](https://icons.getbootstrap.com/icons/union/)

**Comment by VeckoTheGecko** [comment]
Updated icons:
image
image

File: tailwind.config.cjs [hunk 1]
```diff
@@ -4,5 +4,6 @@ module.exports = {
   theme: {
     extend: {},
   },
+  darkMode: "class",
```
Comment (VeckoTheGecko) [inline tailwind.config.cjs:7]: Just wondering, what does this do?

**Review by VeckoTheGecko — CHANGES_REQUESTED** [review]
Thanks @dmlb for this featureset! Great usability progress and welcome features with the filter modes and dark mode toggle.

Still some work to be done with refactoring (namely reducing duplication in filter handlers), but this is a huge step in the right direction. Further refactoring can be done in other PRs.

A couple minor points to address before shipping.

**Comment by VeckoTheGecko** [comment]
Any opinions on using [TypeScript enums](https://www.typescriptlang.org/docs/handbook/enums.html) for categorical options in the codebase rather than strings? No need to push any changes re. them (the types `<string> | <string>` are expressive as is), just curious whether you've encountered them before and whether its something to consider in the codebase in general for some items.

File: src/lib/components/FilterForm.svelte [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+        filterOptions.forEach(option => option.active = false)
+        dispatch("filter", {filterOptions, filterLogic})
+    }
+
+    const onSubmit = () => {
+        dispatch("filter", {filterOptions, filterLogic})
+    }
+</script>
+
+{#if filterOptions}
```
[the remark this answers is not in the document]
Comment (dmlb) [reply src/lib/components/FilterForm.svelte:34]: it protects loading binding all the logic if it were loaded and incorrect props passed, but low concern in such a small codebase I'm guessing
good habits die hard 😆

[the remark this answers is not in the document]
Comment (VeckoTheGecko) [reply src/lib/components/FilterForm.svelte:34]: Ah I see. I think you've thought it through more than I have, happy to leave the decision up to you :)

**Comment by VeckoTheGecko** [comment]
I think that all good now. Let me know if we're good to merge @dmlb and we'll get this shipped

**Comment by dmlb** [comment]
> I think that all good now. Let me know if we're good to merge @dmlb and we'll get this shipped

good to go :+1:

**Review by VeckoTheGecko — APPROVED** [review]

**Pull request merged**